framework/resource/manager.go

```diff
@@ -305,6 +305,83 @@ func (m *Manager) DestroyAll(args ...interface{}) error {
 	return result.Err()
 }
 
+// StatusAll invokes the GetStatus method on all the resources under management.
+// The order in which the status of each resource is queried is
+// non-deterministic, and does rely on any creation order or state of the
+// resource. All the state that was created via Create will be available to the
+// Status callbacks, if any. Resources are not required to have a state to have
+// a status.
+func (m *Manager) StatusAll(args ...interface{}) error {
+	if err := m.Validate(); err != nil {
+		return err
+	}
+
+	mapperArgs, err := m.mapperArgs()
+	if err != nil {
+		return err
+	}
+	for _, arg := range args {
+		mapperArgs = append(mapperArgs, argmapper.Typed(arg))
+	}
+
+	var finalInputs []argmapper.Value
+	// Go through available resources.
+	for _, r := range m.resources {
+		// Create the mapper for status
+		f, err := r.mapperForStatus(nil)
+		if err != nil {
+			return err
+		}
+		mapperArgs = append(mapperArgs,
+			argmapper.ConverterFunc(f),
+			// the status methods should receive the resource state, if any
+			// exists
+			argmapper.Typed(r.State()),
+		)
+
+		// Ensure that our final func is dependent on the marker for
+		// this resource so that it definitely gets called.
+		finalInputs = append(finalInputs, markerValue(r.name))
+	}
+
+	// Create our final target function.
+	finalInputSet, err := argmapper.NewValueSet(finalInputs)
+	if err != nil {
+		return err
+	}
+
+	finalFunc, err := argmapper.BuildFunc(
+		finalInputSet, nil,
+		func(in, out *argmapper.ValueSet) error {
+			// no-op on purpose. This function only exists to set the
+			// required inputs for argmapper to create the correct call
+			// graph.
+			return nil
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	// Call it
+	result := finalFunc.Call(mapperArgs...)
+
+	return result.Err()
+}
+
+// StatusReports returns a slice of the current status of all known resource.
+// Returns a nil slice if no resources are known, or if those resources have no
+// status yet.
+func (m *Manager) StatusReports() []*pb.StatusReport_Resource {
```

Comment:
Should this just be merged into `StatusAll` and have StatusAll return the reports?

Reply:
I thought about this and at first decided to be consistent for `StatusAll` to just preform the work and return an error, but I like the idea of combining them better. I'll make that change.

Reply:
Refactored in 2433ce5, `StatusAll()` now returns a slice of reports and error/nil